### PR TITLE
plugin for catching IllegalArgument errors in specific docker containers

### PR DESCRIPTION
**Pull request by gesinaphillips**
[Issue NIHDEV-157](https://pscdevteam.atlassian.net/jira/software/projects/NIHDEV/boards/1?selectedIssue=NIHDEV-157)

This plugin attempts to locate IllegalArgumentExceptions and determine whether they are in a relevant docker container. Path branches to include generic IllegalArgumentExceptions for thoroughness. This PR also includes some abstraction of logic into a parent RegexDiagnosticPlugin and subsequent revision to the IOError plugin.

As I am testing integrating linting into PSC HuBMAP repos, this repo includes changes to quotations and line breaks as well. Apologies for the noise.

### CODE REVIEW

File: src/ingest-pipeline/airflow/dags/diagnostics/plugins/illegal_argument_plugin.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+from itertools import dropwhile, groupby
+
+from diagnostics.diagnostic_plugin import DiagnosticResult, RegexDiagnosticPlugin
+
+"""
+DOCKER_REGEX locates the section of session.log where both docker containers are started
+and captures everything after that, inclusive of the container start log, as a rough
+way of logging an error taking place in this type of docker container
+"""
+DOCKER_REGEX = r"job collect_ometiff_files.*docker.*job convert_to_pyramid.*docker*.*"
```
Comment (jswelling) [inline src/ingest-pipeline/airflow/dags/diagnostics/plugins/illegal_argument_plugin.py:15]: Hey, I've got an idea.  Can this be generalized to catch *any* docker run?  They all start with 'docker ... run' and end with something like "[job ...] Max memory used".  Your new class for detecting docker errors would be a lot more useful if it was more general.
```

**Review by jswelling — CHANGES_REQUESTED** [review]

File: src/ingest-pipeline/airflow/dags/diagnostics/diagnostic_plugin.py [hunk 2]
```diff
@@ -77,6 +78,17 @@ def diagnose(self) -> DiagnosticResult:
         raise NotImplementedError()
 
 
+class RegexDiagnosticPlugin(DiagnosticPlugin):
+    def __init__(self, **kwargs):
+        self.dir_path = Path(kwargs["local_directory_full_path"])
+
+    @property
+    def get_session_log_path(self):
```
[the remark this answers is not in the document]
Comment (gesinaphillips) [reply src/ingest-pipeline/airflow/dags/diagnostics/diagnostic_plugin.py:86]: Ah thank you!

File: src/ingest-pipeline/airflow/dags/diagnostics/plugins/illegal_argument_plugin.py (hunk 1 above)
Comment (gesinaphillips) [reply src/ingest-pipeline/airflow/dags/diagnostics/plugins/illegal_argument_plugin.py]: Gave this a shot, please see RegexDiagnosticPlugin class.

File: src/ingest-pipeline/airflow/dags/diagnostics/diagnostic_plugin.py [hunk 3]
```diff
@@ -83,11 +84,33 @@ def __init__(self, **kwargs):
...
+        docker_regex = r"(\[job )([^\]]*)(.*docker.*)(\2] Max memory used)"
+        compiled_docker_regex = re.compile(docker_regex, re.DOTALL)
+        docker_match = compiled_docker_regex.search(session_log_data)
+        if not docker_match:
+            return None
+        return docker_match
+
+    def get_named_docker_section(self, process_name):
+        session_log_data = open(self.session_log_path).read()
+        docker_regex = rf"(\[job )({process_name}?(?=\]))(.*docker.*)(\2\])( Max memory used)"
```
Comment (jswelling) [inline src/ingest-pipeline/airflow/dags/diagnostics/diagnostic_plugin.py:107]: Ooh, nice, I didn't know you could combine r and f on a string!

**Review by jswelling — APPROVED** [review]
Much, much better and more flexible!

**Pull request merged**